Feat/handle error transform

tiagralee opened the pull request:
issue #1

Add a common.CustomError
Changed svc_handler.sysl to transform the error type to the CustomerError map
Changed svc_handler.sysl add a new function HandleError to map custom error, if not found return the common http error.
Update the tests for simple and dbendpoints.

andrewemeryanz commented on codegen/tests/simple/servicehandler.go:
```diff
@@ -12,6 +12,12 @@ import (
 )
 
+// *BusinessLogicError error
+var BusinessLogicError = map[string]string{"name": "BusinessLogicError", "http_code": "1001", "http_message": "foo", "http_status": "500"}
```
This needs to be:
```go
var BusinessLogicError CustomError = map[string]string{ ... }
```

so you can use it as an error directly:
```go
var e error = BusinessLogicError
```

andrewemeryanz commented on codegen/tests/simple/simple_gencallback.go:
```diff
@@ -41,3 +37,21 @@ func (c Callback) AddMiddleware(ctx context.Context, r chi.Router) {
 	return "/"
 }
+
+func (c CallbackWithMapError) MapError(ctx context.Context, cause error) *common.HTTPError {
```
`MapError` should take a `ServerError`, not just an `error` to make this more explicit.

You can modify `CreateError` to return a `ServerError` as follows:

```go
func CreateError(ctx context.Context, kind Kind, message string, cause error) ServerError {
	// we may push the error to NR here
	if err := CheckContextTimeout(ctx, message, cause); err != nil {
		return *err
	}

	e, ok := cause.(*ServerError)
	if !ok {
		return ServerError{Kind: kind, Message: message, Cause: cause}
	}

	return *e
}

func CheckContextTimeout(ctx context.Context, message string, cause error) *ServerError {
	if ctx.Err() == context.DeadlineExceeded {
		return &ServerError{Kind: DownstreamTimeoutError, Message: message, Cause: cause}
	}
	return nil
}
```

tiagralee replied:
This CreateError is referenced by other places as well those requires a generic error, do we need to create separate CreateServerError method instead?

andrewemeryanz replied:
Ok. Leave it for now.

andrewemeryanz commented on common/error_handler.go:
```diff
@@ -13,10 +14,30 @@ const (
+func HandleError(ctx context.Context, w http.ResponseWriter, kind Kind, message string, cause error, httpErrorMapper func(context.Context, error) *HTTPError) {
+	err := CreateError(ctx, kind, message, cause)
 	logEntry := GetLogEntryFromContext(ctx)
-	logEntry.Error(err)
```
Can we just log the `ServerError` here instead of the `cause`? The `ServerError` contains the `cause`.

andrewemeryanz reviewed (changes requested):
Couple of changes requested. Thanks.

andrewemeryanz commented on codegen/tests/simple/simple_gencallback.go:
```diff
@@ -41,3 +37,21 @@ func (c Callback) AddMiddleware(ctx context.Context, r chi.Router) {
 }
+
+func (c CallbackWithMapError) MapError(ctx context.Context, cause error) *common.HTTPError {
+	return &common.HTTPError{HTTPCode: 500, Code: "1001", Description: "foo"}
```
This should be passed in from the test rather than hardcoded (to give your tests flexibility). Add a field to the `Callback` and remove `CallbackWithMapError`:
```go
type Callback struct{
    MapError *common.HTTPError
}

func (c CallbackWithMapError) MapError(ctx context.Context, cause error) *common.HTTPError {
    return c.MapError
}
```

then just pass in your error to test:
```go
cb := Callback{ &common.HTTPError{ ... } }
``

gkanz commented on codegen/transforms/svc_types.sysl:
```diff
@@ -309,7 +309,7 @@ CodeGenTransform:
       Comment = "// Code generated by sysl DO NOT EDIT.\n"
 
-      TopLevelDecl = unused | GoStructs(app.types) | GoSetType(app.types) | RequestTypes(app.endpoints, module) | validatorDecl(app.types) |  addDecl(app.types) | lookupDecl(app.types) | GoAlias(app.alias where(HasPrefix(.key, "EXTERNAL_") == false && .key != "Empty")) | GoEmptyType(app.alias where(.key == "Empty"))
+      TopLevelDecl = unused | GoStructs(app.types where(.attrs.patterns == null)) | GoSetType(app.types where(.attrs.patterns == null)) | RequestTypes(app.endpoints, module) | validatorDecl(app.types where(.attrs.patterns == null)) |  addDecl(app.types where(.attrs.patterns == null)) | lookupDecl(app.types where(.attrs.patterns == null)) | GoAlias(app.alias where(HasPrefix(.key, "EXTERNAL_") == f ...
```
This will generate spurious warnings, should use `where("patterns" !in .attrs))` instead of `where(.attrs.patterns == null)`.

!in is missing in Sysl. I have the necessary changes: https://github.com/anz-bank/sysl/tree/add-not-in-binexpr but I need to write the tests.

tiagralee replied:
This is not merged yet, can I leave if for now and create an issue to change it after the sysl update?

gkanz replied:
OK, but you will get complaints about the warnings.

andrewemeryanz replied:
@gkanz Can we merge this in with a `TODO` and revisit after your changes?

gkanz commented on common/error_handler.go:
```diff
@@ -13,10 +14,26 @@ const (
+		switch t := cause.(type) {
+		case CustomError:
+			e := t.HTTPError(ctx)
+			httpError = &e
```
Can this be a single line?

andrewemeryanz replied:
> Can this be a single line?

Yes, if `t.HTTPError` returns a pointer.

gkanz commented on common/error_handler.go:
```diff
@@ -13,10 +14,26 @@ const (
+			httpError = &e
+		default:
+			e := MapError(ctx, err)
+			httpError = &e
```
Can this be a single line?

andrewemeryanz replied:
> Can this be a single line?

No unfortunately. `&MapError(ctx, err)` is invalid Go.

gkanz commented on common/custom_error.go:
```diff
@@ -0,0 +1,34 @@
+	}
+	httpCode := getOrDefault(e, "http_code", "")
+	httpMessage := getOrDefault(e, "http_message", "")
+	return HTTPError{httpStatus, httpCode, httpMessage}
```
Is there a reason we are making a copy of the `HTTPError` here? Could `httpMessage` be large? Would returning a pointer help with:
```Go
e := t.HTTPError(ctx)
httpError = &e
```

andrewemeryanz replied:
>  Could httpMessage be large?

It's unlikely to be of any great length but yes, a pointer can be returned here.

andrewemeryanz commented on core/callback.go:
```diff
@@ -19,7 +18,7 @@ type RestGenCallback interface {
 	Config() validator.Validator
 	// HandleError allows custom HTTP errors to be added to `sys-go` errors
-	HandleError(ctx context.Context, w http.ResponseWriter, kind common.Kind, message string, cause error)
+	MapError(ctx context.Context, cause error) *common.HTTPError
```
Minor request. Can we rename `cause` to `err`. The actual error passed is a wrapped error, not the original cause.

andrewemeryanz commented on core/callback.go:
```diff
@@ -19,7 +18,7 @@ type RestGenCallback interface {
 	// This is returned from the status endpoint
 	Config() validator.Validator
 	// HandleError allows custom HTTP errors to be added to `sys-go` errors
```
Also, can we update this documentation:
```
MapError maps an error to an HTTPError in instances where custom error mapping is 
required. Return nil to perform default error mapping; defined as: 
1. CustomError.HTTPError if the original error is a CustomError, otherwise
2. common.MapError
```

Commit:
Change callback to 'map' error rather than 'handle' error

The current approach that the framework takes to handle internal errors
is to have generated code call custom code (via a callback) and request
that it 'handle' the error. Handling the error involves:
1. Logging the internal error
2. Mapping the internal error into an appropriate http error

Logging an internal error is something that the framework already handles
and, in the general case, mapping an internal error into an http error
is something that could be handled automatically if such a mapping
existed.

This change does the following:
1. Introduce error types (defined in sysl)
2. Change the interaction between the framework and custom code to
support internal error handling.

Internal errors are now 'handled' by the framework. The framework will
automatically log the internal error, then request that custom code
map the error to an http error if desired. If the internal error is
one of the error types defined in sysl (above), then custom code can
return nil (ignoring any custom mapping) and the framework will return
the values as defined in the sysl definition.

An example of an error type defined in sysl is:

!type BusinessLogicError [~error]:
    http_status <: string [value = "500"]
    http_code <: string [value = "1001"]
    http_message <: string [value = "foo"]

Requires: sysl v0.42.0

Closes: #1
